feat(tree): Add `TableSchema` APIs (alpha) (#24579)
microsoft/FluidFramework: .changeset/tabaxi-fellas-inspire.md

---
"fluid-framework": minor
"@fluidframework/tree": minor
"__section": tree
---
TableSchema (alpha) APIs have been added

> Users may now represent dynamic, tabular data using Shared Tree.
> New APIs have been added under a `TableSchema` namespace for creating column, row, and table node schema.

Reviewer: They could do this before. This change provides a standard easy way to do it. I think something liek this would be better:

```suggestion
A `TableSchema` utility has been added to Shared Tree for managing dynamic, tabular data.
This new `TableSchema` namespace contains APIs for creating column, row, and table [node schema](https://fluidframework.com/docs/api/fluid-framework/treenodeschema-typealias).
```

Note: these APIs require the use of [SchemaFactoryAlpha](https://fluidframework.com/docs/api/fluid-framework/schemafactoryalpha-class).

#### Creating a table with default column and row schema

You can craft a table with defaults via `TableSchema.createTable`, without specifying custom column or row schema.

> Note that you will still be required to provide a schema for the cells that will appear in the table.

Author: The heading is more explicit, but I certainly don't see the harm in being more explicit here as well.

```typescript
class Cell extends schemaFactory.object("TableCell", {
	value: schemaFactory.string,
}) {}

class Table extends TableSchema.createTable({
	schemaFactory,
	cell: Cell,
}) {}

const table = new Table({
	columns: [{ id: "column-0" }],

> 	rows: [{ id: "row-0", cells: {} }],

Reviewer: I think this examples should either:

1. Actually, create a cell inline to show the data model for how cells work (put it in the column)

OR

2. Use the methods to specific columns and rows, rather than as inline insertable content. Maybe even have and use a Table.Empty() static to fully hide the details.

Author: I like both.

Author: Actually, for the latter recommendation, is there any straightforward way for the factory functions to tack on static functionality to the _derived_ schema classes?

> });

Author: Yeah, that's fair. I'll update the examples (these same examples live in tests, TSDoc, and here)

```

#### Creating a table with custom column and row schema

> If you need to associate additional data with your rows and columns, you can customize your schema via `TableSchema.createColumn` and `TableSchema.createRow`.

Reviewer: ```suggestion
If you need to associate additional data with your rows or columns, you can customize your row and column schema via `TableSchema.createColumn` and `TableSchema.createRow`.
```

> These schema can then be provided to `TableSchema.createTable`.

Reviewer: ```suggestion
These schema can then be provided to `TableSchema.createTable`:
```

```typescript
class Cell extends schemaFactory.object("TableCell", {
	value: schemaFactory.string,
}) {}

class ColumnProps extends schemaFactory.object("TableColumnProps", {
	// Column label to display.
	label: schemaFactory.string,

> 	// The type of data represented by the cells. Default: string.

Reviewer: I'm a bit confused here. Is the intended implication here that this optional field stores some string describing some concept which the cells in the column "represent", and the default concept (when this field us undefined is described the string "string" which isn't quoted for some reason?

Also its best practice to document fields with doc comments so IntelliSense works for them.

Regardless, I think the same argument I made about not defining a Cell schema as is districts from the example applies here as well. I think some simpler inline schema in the createColumn would be more clear.

Author: The example has been simplified.

	dataType: schemaFactory.optional(schemaFactory.string),
}) {}

class Column extends TableSchema.createColumn({
	schemaFactory,
	props: ColumnProps,
}) {}

class Row extends TableSchema.createRow({
	schemaFactory,
	cell: Cell,
}) {}

class Table extends TableSchema.createTable({
	schemaFactory,
	cell: Cell,
	column: Column,
	row: Row,
}) {}

const table = new Table({
	columns: [
		new Column({ props: { label: "Entry", dataType: "string" } }),
		new Column({ props: { label: "Date", dataType: "date" } }),
		new Column({ props: { label: "Amount", dataType: "number" } }),
	],
	rows: [],
});
```

#### Listening for changes

> Listening for changes to table trees behaves just like it would for any other nodes in a Shared Tree.

Reviewer: I think you would link the relevant APi docs for this.

> ---

Reviewer: None of the code examples use the methods on the table node, which I think are one of the main reasons this library exists and the main way it should be used.

A simple example creating a table, creating a table with three columns, two of the same type, and one of a different type, then inserting some rows with cell and reading and editing a cell using the types from the column would make for a nice simple example.

Since there are no examples of it yet, do we have an API for doing strongly types reading of cells using the column?

Also we might want to show allowing row and cell ids to default to uuids. I know that will cause issue using the methods currently on unhydrated nodes, but I'm working on fixes for that and it might be worth highlighting the problem.

Author: I'll add some more usage examples.

This new `TableSchema` namespace contains APIs for creating column, row, and table [node schema](https://fluidframework.com/docs/api/fluid-framework/treenodeschema-typealias).

Note: these APIs require the use of [SchemaFactoryAlpha](https://fluidframework.com/docs/api/fluid-framework/schemafactoryalpha-class).

> [!WARNING]
> These APIs are in preview and are subject to change.
> Until these APIs have stabilized, we do not recommend using them in production code.
> We reserve the right to make breaking changes to these APIs, including their persisted data format.
> Using these APIs in production code may result in data loss or corruption.

#### Creating a table with default column and row schema

You can craft a table with defaults via `TableSchema.createTable`, without specifying custom column or row schema.
Note that you will still be required to provide a schema for the cells that will appear in the table.

```typescript
class Cell extends schemaFactory.object("TableCell", {
	value: schemaFactory.string,
}) {}

class Table extends TableSchema.createTable({
	schemaFactory,
	cell: Cell,
}) {}

const table = new Table({

> 	columns: [{ id: "column-0" }],

Reviewer: I think it will be common for users to want to declare an unhydrated column node in its own variable so they can then reference it in the table and use it manuliplating cells. We might want to show that pattern somewhere, which could be here.

---
"fluid-framework": minor
"@fluidframework/tree": minor
"__section": tree
---
TableSchema (alpha) APIs have been added

A `TableSchema` utility has been added to Shared Tree for managing dynamic, tabular data.
This new `TableSchema` namespace contains APIs for creating column, row, and table [node schema](https://fluidframework.com/docs/api/fluid-framework/treenodeschema-typealias).

Note: these APIs require the use of [SchemaFactoryAlpha](https://fluidframework.com/docs/api/fluid-framework/schemafactoryalpha-class).

> [!WARNING]
> These APIs are in preview and are subject to change.

> > Until these APIs have stabilized, we do not recommend using them in production code.
> > We reserve the right to make breaking changes to these APIs, including their persisted data format.

Reviewer: As recommended by our linter, "we" isn't super well defined in all the context this message may occur in. I suggest updating this and the comment on the namespace. Maybe something like:

```suggestion
> Until these APIs have stabilized, it is not recommended to use them in production code.
> They may be breaking changes to these APIs and their underlying data format.
```

> Using these APIs in production code may result in data loss or corruption.

#### Creating a table

You can craft a table schema with `TableSchema.table`.

> Note that you will still be required to provide a schema for the cells that will appear in the table.

Reviewer: I think we can make this less wordy:

```suggestion
This includes providing a schema for the cells that will appear in the table:
```

```typescript
class Table extends TableSchema.table({
	schemaFactory,
	cell: schemaFactory.string,
}) {}

const table = new Table({
	columns: [{ id: "column-0" }],
	rows: [{ id: "row-0", cells: { "column-0": "Hello world!" } }],
});
```

#### Creating a table with custom column and row schema

> If you need to associate additional data with your rows or columns, you can customize your row and column schema via `TableSchema.column` and `TableSchema.row`.
> These schema can then be provided to `TableSchema.table`:

Reviewer: ```suggestion
To associate additional data with your rows or columns, generate custom row and column schema using `TableSchema.column` and `TableSchema.row`.
These schema can then be provided to `TableSchema.table`:
```

```typescript
const Cell = schemaFactory.string;

class ColumnProps extends schemaFactory.object("TableColumnProps", {

> 	/** The column label to display. */

Reviewer: I think its more clear that this is an arbitrary example value this way:
```suggestion
	/** A column label to display. */
```

	label: schemaFactory.string,
}) {}

class Column extends TableSchema.column({
	schemaFactory,
	props: ColumnProps,
}) {}

class Row extends TableSchema.row({
	schemaFactory,
	cell: Cell,
}) {}

class Table extends TableSchema.table({
	schemaFactory,
	cell: Cell,
	column: Column,

> 	row: Row,

Reviewer: Code like this makes me wonder if it would make sense to capitalize the properties which are passed schema so you could just do:

```
	Cell,
	Column,
	Row,
```

Author: Yeah, I've been wondering the same. Our SharedTree examples leverage `PascalCase` for derived schema classes _consistently_, yes? Or, at least, that is generally the guidance we use / recommend? If so, I'm good with such a rename.

Reviewer: I try and recommend PascalCase for schema everywhere, but we don't often have schema as members of objects where this pattern would make sense. We do have a few though, and they don't follow this pattern for example the schema in TreeViewConfiguration is called schema not Schema.

Another example, SchemaFactory.string is lowercase.

I think I'm going to side against my own idea here since it would be really odd in cases where the schema are AlowedTypes unions. Also, I think we want to encourage people to use more semantic names than Cell, Column, and Row.

Reviewer 2: I think the shorthand would be slick, but only applicable sometimes. I agree with Craig.

}) {}

const table = new Table({
	columns: [

> 		new Column({ props: { label: "Entry" } }),

Author: Good call! I'll update our examples :)

		new Column({ props: { label: "Date" } }),
		new Column({ props: { label: "Amount" } }),
	],
	rows: [],
});
```

#### Interacting with the table

Table trees created using `TableSchema` offer various APIs to make working with tabular data easy.
These include:

- Insertion and removal of columns, rows, and cells.
- Cell access by column/row.

```typescript
// Create an empty table
const table = Table.empty();

const column0 = new Column({
	props: { label: "Column 0" },
});

// Append a column to the end of the table.
table.insertColumn({
	column: column0,
});

const rows = [
	new Row({ cells: { } }),
	new Row({ cells: { } }),
];

// Insert rows at the beginning of the table.
table.insertRows({
	index: 0,
	rows,
});

// Set cell at row 0, column 0.
table.setCell({
	key: {
		column: column0,
		row: rows[0],
	},
	cell: "Hello",
});

// Set cell at row 1, column 0.
table.setCell({
	key: {
		column: column0,
		row: rows[1],
	},
	cell: "World",
});

// Remove the first row.
// Note: this will also remove the row's cell.
table.removeRow(rows[0]);

// Remove the column.

> // Note: this will *not* remove the remaining cell under this column.

Reviewer: I think it would be useful to have an example that allows for more than one column type, or at least one where the schema for the column and its use are not exactly the same semantically (for example: having a string column schema which you use for a description instead of using a description column for a description).

Using the same schema for two different columns would be a good way to show off this separation.

If its a bit much to include here, maybe just now the possibility and link the example app? I assume it does this kind of thing?

Author: I like the idea of linking to Nick's demo app. Though, because these APIs aren't published yet, it doesn't yet use them. Might be something to add when we promote these APIs farther?

An alternative might be to link to our internal "example" app, but I don't think we generally point users to those.

Author: I'll add a test case for this kind of scenario. Not sure if we'll want to add it here or not, but a unit test capturing the scenario seems valuable either way.

Reviewer: You can link to the unit test. I think that would be fine for changeset, but longer lived docs can link Nicks example once its ready.

Author: I added a test and noted the scenario in the changeset

});

// Listen for any changes to the table and its children.
// The "treeChanged" event will fire when the `table` node or any of its descendants change.
Tree.on(table, "treeChanged", () => {
	// Respond to the change.
});
```

If you need more granular eventing to meet your performance needs, that is possible as well.
For example, if you wish to know when the table's list of rows changes, you could do the following:

```typescript
class Cell extends schemaFactory.object("TableCell", {
	value: schemaFactory.string,
}) {}

class Table extends TableSchema.table({
	schemaFactory,
	cell: Cell,
}) {}

const table = new Table({
	columns: [{ id: "column-0" }],
	rows: [{ id: "row-0", cells: {} }],
});

// Listen for any changes to the list of rows.
// The "nodeChanged" event will fire only when the `rows` node itself changes (i.e., its own properties change).
// In this case, the event will fire when a row is added or removed, or the order of the list is changed.
// But it won't fire when a row's properties change, or when the row's cells change, etc.
Tree.on(table.rows, "nodeChanged", () => {
	// Respond to the change.
});
```

#### Limitations

##### Orphaned cells

> Note: for now it is possible for table cells to become "orphaned."

Reviewer: ```suggestion
Cells in the table can become "orphaned."
```

That is, it is possible to enter a state where one or more rows contain cells with no corresponding column.

> To help avoid this situation, you can manually remove corresponding cells when removing columns.

Reviewer: ```suggestion
To reduce the chance of this, you can manually remove corresponding cells when removing columns.
```

> [!WARNING]
> Note that even with the above precaution, it is possible to enter such an orphaned cell state via the merging of edits.
> For example: one client might add a row while another concurrently removes a column, orphaning the cell where the column and row intersected.

##### Column-specific cell type constraints

Currently, the table APIs do not offer any ability to express cell-typing constraints on a per-column basis.
A cell type can be defined for *all* cells in the table, uniformly.

It is possible, however, to extend the APIs to add some level of runtime validation to enforce stricter cell typing at the column level.

> A simplified example of such a pattern can be found in `tableSchema.spec.ts` in a test called "Using column props for runtime type constraints".

Reviewer: I think you should link to the actual source file on github, as readers of this might have a hard time getting from someplace like the release notes, or the changelog from the FF package.

Author: Example removed. Will follow up separately.
